crates/publish/src/oci/client.rs

```diff
@@ -0,0 +1,250 @@
+//! Client for distributing Spin applications using OCI registries.
+
+use anyhow::{bail, Context, Result};
+use docker_credential::{CredentialRetrievalError, DockerCredential};
+use oci_distribution::{
+    client::{Config, ImageLayer},
+    manifest::OciImageManifest,
+    secrets::RegistryAuth,
+    Reference,
+};
+use spin_app::locked::{ContentPath, ContentRef};
+use spin_loader::oci::cache::Cache;
+use spin_manifest::Application;
+use tokio::fs;
+use walkdir::WalkDir;
+
+use std::path::{Path, PathBuf};
+
+// TODO: the media types for application, wasm module, and data layer are not final.
+const SPIN_APPLICATION_MEDIA_TYPE: &str = "application/vnd.fermyon.spin.application.v1+config";
+const WASM_LAYER_MEDIA_TYPE: &str = "application/vnd.wasm.content.layer.v1+wasm";
```

Comment:
Just a drive-by-comment: I think media type can be added to image-spec, see discussion here: https://github.com/opencontainers/image-spec/pull/964#issuecomment-1278087823

Reply:
Note that this is the media type of a layer, not of the top-level manifest or of the config.

In this scenario, the _artifact_ is a Spin application, and it contains Wasm layers.
We could use the upcoming referrer API, but that's not something stable AFAIK.

Reply:
Thanks for the clarification. SGTM :)